builder/tart/vnc.go

```diff
@@ -0,0 +1,285 @@
+                    return true;
+            }
+        }
+    }
+
+    return false;
+}
+*/
+import "C"
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/packer-plugin-sdk/bootcommand"
+	"github.com/mitchellh/go-vnc"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"bytes"
+	"compress/zlib"
+	"encoding/binary"
+
+	"image"
+	"image/color"
+
+	"unsafe"
+)
+
+type customDriver struct {
+	vncClient            *vnc.ClientConn
+	serverMessageChannel chan vnc.ServerMessage
+	config               *Config
+	vncDriver            bootcommand.BCDriver
+	keyInterval          time.Duration
+	ctx                  context.Context
+	frameBuffer          *image.RGBA
+	waitString           strings.Builder
+}
+
+func newCustomDriver(vncClient *vnc.ClientConn,
+	serverMessageChannel chan vnc.ServerMessage,
+	config *Config,
+	ctx context.Context) *customDriver {
+
+	// Resolve key interval manually so we can accurately report it back
+	keyInterval := bootcommand.PackerKeyDefault
+	if delay, err := time.ParseDuration(os.Getenv(bootcommand.PackerKeyEnv)); err == nil {
+		keyInterval = delay
+	}
+	if config.BootKeyInterval > time.Duration(0) {
+		keyInterval = config.BootKeyInterval
+	}
+
+	w, h := int(vncClient.FrameBufferWidth), int(vncClient.FrameBufferHeight)
+
+	d := &customDriver{
+		vncClient:            vncClient,
+		serverMessageChannel: serverMessageChannel,
+		config:               config,
+		vncDriver:            bootcommand.NewVNCDriver(vncClient, keyInterval),
+		keyInterval:          keyInterval,
+		ctx:                  ctx,
+		frameBuffer:          image.NewRGBA(image.Rect(0, 0, w, h)),
+	}
+
+	return d
+}
+
+func (d *customDriver) KeyInterval() time.Duration {
+	return d.keyInterval
+}
+
+const WaitForStringStart uint32 = 0xE000
+const WaitForStringEnd uint32 = 0xE0001
+
+func (d *customDriver) SendKey(key rune, action bootcommand.KeyAction) error {
+	switch uint32(key) {
+	case WaitForStringStart:
+		d.waitString.Grow(1)
+		return nil
+	case WaitForStringEnd:
+		waitString := d.waitString.String()
+		d.waitString.Reset()
+
+		waitStringCStr := C.CString(waitString)
+		defer C.free(unsafe.Pointer(waitStringCStr))
+
+		for {
+			fmt.Fprintf(os.Stderr, "Waiting for '%s'...\n", waitString)
+			if err := d.WaitForFramebufferUpdate(); err != nil {
+				return err
+			}
+
+			if C.recognizeTextInFramebuffer(waitStringCStr,
+				unsafe.Pointer(&d.frameBuffer.Pix[0]),
+				C.int(d.frameBuffer.Bounds().Dx()),
+				C.int(d.frameBuffer.Bounds().Dy())) {
+				break
+			}
+		}
+
+		return nil
+	default:
+		if d.waitString.Cap() > 0 {
+			d.waitString.WriteRune(key)
+			return nil
+		} else {
+			return d.vncDriver.SendKey(key, action)
+		}
+	}
+}
+
+func (d *customDriver) SendSpecial(special string, action bootcommand.KeyAction) error {
+	return d.vncDriver.SendSpecial(special, action)
+}
+
+func (d *customDriver) Flush() error {
+	return d.vncDriver.Flush()
+}
+
+func (d *customDriver) WaitForFramebufferUpdate() error {
+
+	incremental := true
```

Comment:
This seems to be always `true`.

Reply:
Ack, leftover from when I had some logic to request full frames.

Reply:
Done.